sdks/java/core/src/main/java/org/apache/beam/sdk/schemas/Schema.java

```diff
@@ -950,6 +1007,297 @@ public int hashCode() {
+
+      public Builder setInt32Option(String fieldName, Integer value) {
+        setOption(fieldName, FieldType.INT32, value);
+        return this;
+      }
+
+      public Builder setInt64Option(String fieldName, Long value) {
+        setOption(fieldName, FieldType.INT64, value);
+        return this;
+      }
+
+      public Builder setDecimalOption(String fieldName, BigDecimal value) {
+        setOption(fieldName, FieldType.DECIMAL, value);
+        return this;
+      }
+
+      public Builder setFloatOption(String fieldName, Float value) {
+        setOption(fieldName, FieldType.FLOAT, value);
+        return this;
+      }
+
+      public Builder setDoubleOption(String fieldName, Double value) {
+        setOption(fieldName, FieldType.DOUBLE, value);
+        return this;
+      }
+
+      public Builder setStringOption(String fieldName, String value) {
+        setOption(fieldName, FieldType.STRING, value);
+        return this;
+      }
+
+      public Builder setDateTimeOption(String fieldName, ReadableDateTime value) {
+        setOption(fieldName, FieldType.DATETIME, value);
+        return this;
+      }
+
+      public Builder setBooleanOption(String fieldName, Boolean value) {
+        setOption(fieldName, FieldType.BOOLEAN, value);
+        return this;
+      }
+
+      public <V> Builder setArrayOption(String fieldName, FieldType arrayType, List<V> value) {
+        setOption(fieldName, arrayType, value);
+        return this;
+      }
+
+      public <K, V> Builder setMapOption(String fieldName, FieldType mapType, Map<K, V> value) {
+        setOption(fieldName, mapType, value);
+        return this;
+      }
+
+      public Builder setRowOption(String fieldName, Row value) {
+        setOption(fieldName, FieldType.row(value.getSchema()), value);
+        return this;
+      }
+
+      public Builder setOption(String fieldName, FieldType fieldType, Object value) {
+        options.put(
+            fieldName, new Option(fieldType, ValueUtils.verify(value, fieldType, fieldName)));
+        return this;
+      }
+
+      public Options build() {
+        return new Options(this.options);
+      }
+
+      public Builder addOptions(Options options) {
+        this.options.putAll(options.options);
+        return this;
+      }
+    }
+
+    Options(Map<String, Option> options) {
+      this.options = options;
+    }
+
+    Options() {
+      this.options = new HashMap<>();
+    }
+
+    Options.Builder toBuilder() {
+      return new Builder(this.options);
+    }
+
+    public static Options.Builder builder() {
+      return new Builder();
+    }
+
+    public static Options none() {
+      return new Options();
+    }
+
+    @SuppressWarnings("TypeParameterUnusedInFormals")
+    private <T> T getValueOrNull(String fieldName) {
+      Option option = options.get(fieldName);
+      if (option != null) {
+        return option.getValue();
+      }
+      return null;
+    }
+
+    @Nullable
+    Byte getByte(String fieldName) {
```

Comment:
This is a lot of getters. I would intead make a single generic getter, like this:

public <T> T get(String fieldName, Class<T> clazz) {
  return (T) ...;
}

then a caller can write

byte[] bytes = option.get(byte[].class);

Reply:
Shouldn't I rename getValueOrNull to get make 

>     @SuppressWarnings("TypeParameterUnusedInFormals")
>     private <T> T getValueOrNull(String fieldName) {

And drop the typed setters as well?

Reply:
Replaced all the getters. Kept all the setters though.